### ISSUE TITLE
Graph classification with edge features - adapt top-k pool examples

### ISSUE THREAD

**Issue by MarionMeyers** [rusty1s/pytorch_geometric#354]
## ❓ Questions & Help

Hello again, 

First of all, many thanks for this library, I find it very complete and useful. 

I am currently trying to perform graph classification on networks that have 2 features on each edges. I have been adapting the Enzyme Top-K example to my case and it is working well. However, when I further looked into the code, I realized that I don't think the edges featues are taken into account. 

I have adapted the first part of the forward method by adding the edge_attr variable in this way :
`
 def forward(self, data):

        x, edge_index, edge_attr, batch = data.x, data.edge_index, data.edge_attr, data.batch

        x = F.relu(self.conv1(x, edge_index))
        x, edge_index, edge_attr, batch, _ = self.pool1(x, edge_index, edge_attr, batch)
        x1 = torch.cat([gmp(x, batch), gap(x, batch)], dim=1)
`
However, I don't think my understanding of the code is good enough for me to understand if the edge feature are actually used for training or not. 

Could you help me on this matter? And if the edges features are not taken into account, how can I go about making this happen? 

Thank you for your time, 
Marion

**Comment by rusty1s** [comment]
Hi,
The `topk` method finds the top k nodes in a graph based on its node features, and removes all other nodes from the graph, hence the graph gets coarsened, and edge indices and edge attributes get filtered out. Therefore, `self.pool1` returns new filtered `x`, `edge_index`, `edge_attr`, and `batch` tensors.

Please note that the topk method does not make use of edge features to find the best nodes. This process is simply based on the node features. Instead, you need to use a GNN which takes edge features into account, like `NNConv` or `EdgeConv`.

**Comment by Dahabium** [comment]
Dear @rusty1s , thanks for this great library! 
First of all, I would like to mention that I dont have a strong deep learning background.
My question is related to previous. I have a binary graph classification problem with 6 features on nodes and 1 feature on edges. My graph (in each data object) could be disconnected - i.e. multiple graphs that are related to each other but not necessarly connected. Also, my nodes dont have a position atribute 'pos'. My edges are directed. 

My first attempt was to try NNConv (I have tried to adapt mnist_nn_conv.py).
The code is: 
`    def __init__(self):
        super(Net, self).__init__()
        numNodesFeatures = len(graph_dataset[0].x[0])
        numClasses = 2
        nn1 = nn.Sequential(nn.Linear(1,12), nn.ReLU(), nn.Linear(12, numNodesFeatures * 32 ))
        self.conv1 = NNConv(numNodesFeatures, 32, nn1, aggr='mean')
        nn2 = nn.Sequential(nn.Linear(1, 12), nn.ReLU(), nn.Linear(12, 32*64))
        self.conv2 = NNConv(32, 64, nn2, aggr='mean')
        self.fc1 = torch.nn.Linear(64, 128)
        self.fc2 = torch.nn.Linear(128, numClasses)
        
    def forward(self, data):
        data.x = F.elu( self.conv1( data.x, data.edge_index, data.edge_attr ) )
        weight = normalized_cut(data.edge_index, data.edge_attr, int(data.num_nodes) )
        cluster = graclus(data.edge_index, weight, int(data.num_nodes))
        data = max_pool(cluster, data)
        data.x = F.elu(self.conv2(data.x, data.edge_index, data.edge_attr))
        weight = normalized_cut(data.edge_index, data.edge_attr, int(data.num_nodes))
        cluster = graclus(data.edge_index, weight, int(data.num_nodes))
        x, batch = max_pool_x(cluster, data.x, data.batch)
        x = global_mean_pool(x, batch)
        x = F.elu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        return F.log_softmax(self.fc2(x), dim=1)    `

I had to modify the normalized_2d_cut to make it work without pos attributes - (is it done correctly?)
My classification results are not improving consistently when training, (they fluctuate from 55-90% test accuracy) (but it also could be because of an inconsistant/incomplete dataset, which may just have small graphs that give not enough confidence to classify them. I also might be giving too large graphs to the network (In the paper it says that NNConv doeset generalize well on large graphs, but not sure about that)
Could you please suggest some guidelines when building the architecture (deciding on the size of convolutional, fully connected and pooling layers? )
I'm planning to try SplineConv, EdgeConv and maybe GCN conv in near future.

Thank you very much in advance!
Really happy to see the community growing.

**Comment by rusty1s** [comment]
Fluctuations in accuracy usually are stemming from small datasets (like in the Graph Kernel benchmark datasets), but lower learning rates can help. Generally, is is quite hard to give any tipps about hyperparameters, but I always suggest to start as simple as possible and try to improve the results further. That means, I would generally omit pooling, and create an MLP baseline to compare against.
